doc/reference/trd-syscalls.md

```diff
@@ -0,0 +1,715 @@
+different system calls need to pass different amounts of data. A
+command that requests a 64-bit timestamp, for example, needs its
+success to return a `u64`, but its failure can return nothing. In
+contrast, a system call that passes a pointer into the kernel may have
+a simple success return value but requires a failure with one 32-bit
+value so the pointer can be passed back.
+
+Every system call MUST return only one failure and only one success
+variant. Different system calls may use different failure and success
+variants, but any specific system call returns exactly one of each. If an
+operation might have different success return variants or failure return
+variants, then it should be split into multiple system calls.
+
+This requirement of a single failure variant and a single success variant is to simplify
+userspace implementations and preclude them from having to handle many different cases.
+The presence of many difference cases suggests that the operation should be split up --
+there is non-determinism in its execution or its meaning is overloaded. It also fits
+well with Rust's `Result` type.
+
+All 32-bit values not specified for `r0` in the above table are reserved.
+Reserved `r0` values MAY be used by a future TRD and MUST NOT be returned by the
+kernel unless specified in a TRD. Therefore, for future compatibility, userspace
+code MUST tolerate `r0` values that it does not recognize.
+
+3.3 Error Codes
+---------------------------------
+
+All system call failures return an error code. These error codes are a superset of
+kernel error codes. They include all kernel error codes so errors from calls
+on kernel HILs can be easily mapped to userspace system calls when suitable. There
+are additional error codes to include errors related to userspace.
+
+| Value | Error Code  | Meaning                                                                                 |
+|-------|-------------|-----------------------------------------------------------------------------------------|
+| 1     | FAIL        | General failure condition: no further information available.                            |
+| 2     | BUSY        | The driver or kernel is busy: retry later.                                              |
+| 3     | ALREADY     | This operation is already ongoing can cannot be executed more times in parallel.        |
+| 4     | OFF         | This subsystem is powered off and must be turned on before issuing operations.          |
+| 5     | RESERVE     | Making this call requires some form of prior reservation, which has not been performed. |
+| 6     | INVALID     | One of the parameters passed to the operation was invalid.                              |
+| 7     | SIZE        | The size specified is too large or too small.                                           |
+| 8     | CANCEL      | The operation was actively cancelled by a call to a cancel() method or function.        |
```

Comment:
#2431

Reply:
These aren't transactions, just operations.

Yes, we do want CANCEL to only be returned in response to an active cancellation by the client. Its presence is important so that there's a meaningful indication of why an asynchronous operation didn't complete. If other call paths can cancel a client's operation (e.g., the kernel decides to do so), then this should probably live in a different `Error` value. E.g., POSIX has `EINTR` for when the kernel has to back out of an operation. But it's important to note this generally only happens when there are signal handlers installed by a process, i.e., the kernel can't just start arbitrarily cancelling/interrupting.

But suppose, for example, the kernel wants to stop an outstanding operation so it can power down a peripheral. I'd argue this should be an `Error::OFF`.